File: app/src/main/java/com/jonasgerdes/stoppelmap/usecases/schedule/EventAdapter.java
package com.jonasgerdes.stoppelmap.usecases.schedule;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.jonasgerdes.stoppelmap.R;
import com.jonasgerdes.stoppelmap.model.map.MapEntity;
import com.jonasgerdes.stoppelmap.model.schedule.Event;

import java.util.List;

import io.realm.RealmChangeListener;
import io.realm.RealmResults;

/**
 * Created by Jonas on 04.08.2016.
 */
public class EventAdapter extends RecyclerView.Adapter<EventHolder> {

    interface EventActionListener {
        void onFacebookAction(String url);

        void onLocationAction(MapEntity location);
    }

    private List<Event> mEvents;
    private boolean mHideLocationButton;
    private EventActionListener mActionListener;

    public EventAdapter(RealmResults<Event> events) {
        mEvents = events;
        events.addChangeListener(new RealmChangeListener<RealmResults<Event>>() {
            @Override
            public void onChange(RealmResults<Event> element) {
                notifyDataSetChanged();
            }
        });
    }

    @Override
    public EventHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.schedule_event_list_item, parent, false);
        return new EventHolder(view);
    }

    @Override
    public void onBindViewHolder(EventHolder holder, int position) {
        final Event event = mEvents.get(position);
        holder.onBind(event);
        holder.mFacebookButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mActionListener != null) {
                    mActionListener.onFacebookAction(event.getFacebookUrl());
                }
            }
        });
        holder.mLocationButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mActionListener != null) {
                    mActionListener.onLocationAction(event.getLocation());
                }
            }
        });
        if (mHideLocationButton) {
            holder.mLocationButton.setVisibility(View.GONE);
            holder.mLocations.setVisibility(View.GONE);
        }
    }

    @Override
    public int getItemCount() {
        return mEvents.size();
    }

    public void setActionListener(EventActionListener actionListener) {
        mActionListener = actionListener;
    }

    public void setHideLocationButton(boolean hideLocationButton) {
        mHideLocationButton = hideLocationButton;
    }
}

File: app/src/main/java/com/jonasgerdes/stoppelmap/usecase/event/overview/view/event_day/EventAdapter.kt
package com.jonasgerdes.stoppelmap.usecase.event.overview.view.event_day

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.jonasgerdes.stoppelmap.R
import com.jonasgerdes.stoppelmap.model.entity.map.MapEntity
import com.jonasgerdes.stoppelmap.model.events.Event
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject
import kotlinx.android.synthetic.main.event_event_card.view.*
import org.jetbrains.anko.sdk25.coroutines.onClick

/**
 * @author Jonas Gerdes <dev@jonasgerdes.com>
 * @since 09.08.17
 */
class EventAdapter : RecyclerView.Adapter<EventHolder>() {

    private var eventList: List<Event> = ArrayList()
    private val openLocationSubject = BehaviorSubject.create<MapEntity>()
    var events
        get() = eventList
        set(value) {
            eventList = value
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): EventHolder {
        val view = LayoutInflater.from(parent?.context).inflate(
                R.layout.event_event_card, parent, false)
        return EventHolder(view)
    }

    override fun onBindViewHolder(holder: EventHolder, position: Int) {
        holder.onBind(eventList[position])
        holder.itemView.locationButton.onClick {
            eventList[holder.adapterPosition].mapEntity?.let {
                openLocationSubject.onNext(it)
            }
        }
    }

    override fun getItemCount(): Int {
        return eventList.size
    }

    fun locationClicks(): Observable<MapEntity> {
        return openLocationSubject.hide()
    }
}